src/main/java/org/elasticsearch/cluster/metadata/SnapshotMetaData.java

```diff
@@ -330,7 +332,16 @@ public Entry snapshot(SnapshotId snapshotId) {
         return null;
     }
 
+    @Override
+    public MetaData.Custom readFrom(StreamInput in) throws IOException {
+        return FACTORY.readFrom(in);
```

Comment:
We have two contexts in which objects are getting deserialized. One is static context when an object doesn't exist and another is dynamic context when we have object in the previous state and applying diffs. These dual contexts are causing duplication of deserialization methods. It could be avoided by replacing these factories with prototypes, but this will break backward compatibility too much. My hope is still to backport it to 1.6. So, if it's ok with you I would rather convert everything to prototypes as a separate PR that will go only to 2.0.